packages/flutter_test/lib/src/controller.dart

```diff
@@ -527,12 +536,94 @@ abstract class WidgetController {
     int buttons = kPrimaryButton,
     double touchSlopX = kDragSlopDefault,
     double touchSlopY = kDragSlopDefault,
+    Duration duration,
+    double frequency = 60.0,
   }) {
-    assert(kDragSlopDefault > kTouchSlop);
+    assert(touchSlopX > kTouchSlop);
+    assert(touchSlopY > kTouchSlop);
```

Comment:
The `touchSlop` logic seems to be much more sophisticated than the `scroll` logic in `flutter_driver`, and it could generate more input events than `duration * frequency`. Shall we just create a separate `drag` function that doesn't have `touchSlop`, but use `duration` and `frequency` to decide how many events to be generated?

Reply:
`touchSlop` comes in as a bug fix in #26438. I'm feeling it strange too but I don't have good enough reason to change it. I'm expecting in the duration-frequency one because `offset/(duration * frequency)` is very small (if the frequency and duration is reasonable), it will always go to single event for each individual input. However if a user is abusing these numbers (like duration * frequency gives 1), I don't want to trigger that bug, so I'm using this as a fall back.

Reply:
As for driver, I feel it's just we are being lucky not having gone into that bug.

Reply:
Double checked  #26438, without this `touchSlop` the widgets that are having issue is the cupertino style switcher. I guess scrolling is not when there's problem. In `flutter_driver`, there isn't a `drag` but always scrolling. My understanding is `flutter_driver` has not reached the stage of being able to do all operations and no one has used `flutter_driver` to test dragging a cupertino switch. However, changing the timed version to `scroll` doesn't sounds good to me because the operation is not always scrolling either (like the one in `macrobenchmarks/picture_cache`).

Reply:
I'm a little inclined to keep it simple until we've seen actual abuses (like `duration * frequency` gives 1) that require us to add such complex logic. Otherwise, we risk of making many of our tests potentially hard to reason about without actually helping anything. Specifically, since we're going to use this in places where `FlutterDriver.scroll` was used, I felt just matching `FlutterDriver.scroll` input events would be good enough to not trigger any new bugs.

Additionally, the old `touchSlop` logic seems to be specifically designed for `duration == null` case and breaks a single move event into 2 move events. Using such logic to break n events into 2n events looks unnecessary and unintended (suppose the offset is really large so each event's offset is large enough to be separated).

Reply:
So I moved the timed part to `slowDrag` and `slowDragFrom`. PTAL

Reply:
I also added a unit test for `timedDrag`. One more thing that's worth mentioning (already in a comment), the PointerUpEvent and its corresponding PointerData doesn't provide `button` field.